libmikmod/playercode/mplayer.c

```diff
@@ -2272,6 +2276,276 @@ static int DoOktArp(UWORD tick, UWORD flags, MP_CONTROL *a, MODULE *mod, SWORD c
 	return 0;
 }
 
+/*========== Farandole effects */
+
+static void DoFarTonePorta(MP_CONTROL* a)
+{
+	if (!a->main.fadevol)
+		a->main.kick = (a->main.kick == KICK_NOTE) ? KICK_NOTE : KICK_KEYOFF;
+	else
+		a->main.kick = (a->main.kick == KICK_NOTE) ? KICK_ENV : KICK_ABSENT;
+
+	a->farcurrentvalue += a->fartoneportaspeed;
+
+	/* Have we reached our note */
+	BOOL reachedNote = (a->fartoneportaspeed > 0) ? (a->farcurrentvalue >> 16) >= a->wantedperiod : (a->farcurrentvalue >> 16) <= a->wantedperiod;
+	if (reachedNote)
+	{
+		/* Stop the porta and set the periods to the reached note */
+		a->tmpperiod = a->main.period = a->wantedperiod;
+		a->fartoneportarunning = 0;
+	}
+	else
+	{
+		/* Do the porta */
+		a->tmpperiod = a->main.period = (UWORD)(a->farcurrentvalue >> 16);
+	}
+
+	a->ownper = 1;
+}
+
+/* Find tempo factor */
+static SLONG GetFARTempoFactor()
+{
+	return farcurtempo == 0 ? 256 : (128 / farcurtempo);
+}
+
+/* Set the right speed and BPM for Farandole modules */
+static void SetFARTempo(MODULE* mod)
+{
+	/* According to the Farandole document, the tempo of the song is 32/tempo notes per second.
+	   So if we set tempo to 1, we will get 32 notes per second. We then need to translate this
```

Comment:
This comment seems a little excessively long...

Reply:
I know this comment is long. I just want to describe how Farandole uses speed, since it is so different than all other players.